File: content/dota_addons/training_polygon/panorama/scripts/custom_game/announce_message.js
"use strict";
// Place at: content/<addon>/panorama/scripts/custom_game/announcer/announcer.js
// Panel this is attached to must contain #AnnouncerContainer / #AnnouncerBox / #AnnouncerLabel (see announcer.xml)

(function () {

	var container = $.GetContextPanel();
	var label = $("#AnnouncerLabel");

	var hideTimer = null;

	// ---- core display logic -------------------------------------------------

	function ClearTimer() {
		if (hideTimer !== null) {
			$.CancelScheduled(hideTimer);
			hideTimer = null;
		}
	}

	// duration: number of seconds, or <= 0 / undefined for "stays until HideMessage() is called"
	function ShowMessage(message, duration) {
		ClearTimer();

		label.text = $.Localize(message);

		container.RemoveClass("Hidden");
		// one frame delay so the transition from opacity:0 actually plays instead of snapping
		$.Schedule(0, function () {
			container.AddClass("AnnouncerVisible");
		});

		if (typeof duration === "number" && duration > 0) {
			hideTimer = $.Schedule(duration, HideMessage);
		}
	}

	function HideMessage() {
		ClearTimer();
		container.RemoveClass("AnnouncerVisible");
		// wait for the fade-out transition before fully collapsing the panel
		$.Schedule(0.25, function () {
			container.AddClass("Hidden");
		});
	}

	// ---- inbound: server Lua -> this client ----------------------------------
	// Fired by Announcer:ShowMessage()/:HideMessage() on the server via
	// CustomGameEventManager:Send_ServerToAllClients / Send_ServerToPlayer

	GameEvents.Subscribe("announcer_show_message", function (data) {
		ShowMessage(data.message, data.duration);
	});

	GameEvents.Subscribe("announcer_hide_message", function (data) {
		HideMessage();
	});

	// ---- outbound: JS -> server Lua singleton --------------------------------
	// Use these when the trigger originates client-side (e.g. a button in another
	// panel) but you still want the Lua Announcer class to own/broadcast it.

	function RequestShowFromServer(message, duration) {
		GameEvents.SendCustomGameEventToServer("announcer_request_show", {
			message: message,
			duration: (typeof duration === "number") ? duration : -1
		});
	}

	function RequestHideFromServer() {
		GameEvents.SendCustomGameEventToServer("announcer_request_hide", {});
	}

	// ---- public API exposed to any other panel's JS --------------------------
	// e.g. from another script: GameUI.CustomUIConfig().Announcer.ShowMessage("GG", 5);

	GameUI.CustomUIConfig().Announcer = {
		// Goes through the server, so it's consistent/broadcastable and Lua stays authoritative.
		ShowMessage: RequestShowFromServer,
		HideMessage: RequestHideFromServer,

		// Updates only this client's panel directly, no round trip to server.
		// Useful for purely-local UI feedback that shouldn't touch game state.
		ShowMessageLocal: ShowMessage,
		HideMessageLocal: HideMessage
	};

})();

File: game/dota_addons/training_polygon/scripts/vscripts/libraries/announce_message.lua
--[[
	class for showing announce messages at the top of the screen, launchable
	from server Lua directly or from client JS (through custom game events)

	Place at: game/scripts/vscripts/announcer.lua
	Require it once from your addon_game_mode.lua:

		require('announcer')

	Usage from anywhere in server Lua:
		announcer:Show({message="Roshan has spawned!", duration=5})   -- auto-hides after 5s
		announcer:Show({message="Draft phase"})                       -- stays until Hide() is called
		announcer:Hide({})

	Usage from client JS (see announcer.js):
		GameEvents.SendCustomGameEventToServer("announcer_request_show", {message="GG", duration=5})
		GameEvents.SendCustomGameEventToServer("announcer_request_hide", {})
]]

if announcer == nil then
	announcer = class({})
end

function announcer:Init()
	CustomGameEventManager:RegisterListener("announcer_request_show", Dynamic_Wrap(announcer, "Show"))
	CustomGameEventManager:RegisterListener("announcer_request_hide", Dynamic_Wrap(announcer, "Hide"))
end

-- args.message: string to display
-- args.duration: seconds before auto-hide, omit/<=0 to stay until Hide() is called
function announcer:Show(args)
	CustomGameEventManager:Send_ServerToAllClients("announcer_show_message", {
		message = args['message'],
		duration = tonumber(args['duration']) or -1
	})
end

function announcer:Hide(args)
	CustomGameEventManager:Send_ServerToAllClients("announcer_hide_message", {})
end

announcer:Init()